### PR TITLE
[minor_change] Add support for Template BD DHCP Policy configuration in ndo version 4+ with ndo_schema_template_bd_dhcp_policy

### LINKED ISSUE #495: Adding DHCP policy on the bridge domain using the module mso_schema_template_bd_dhcp_policy failure
### Description
* Cannot associate a DHCP relay policy with a BD (DHCP relay policy has been already configured). 

### Affected Module Name(s):
* mso_schema_template_bd_dhcp_policy

### MSO version and MSO Platform
ND: 3.0.1i
NDO: 4.2.3k

### APIC version and APIC Platform for Site Level Resources
APIC: 5.3.2c

### Collection versions
* cisco.mso 2.8

### Output/ Error message
task path: /home/cisco/ansible-ndo/ndo-playbooks/0001_add_bd_dhcp_config.yml:9
....
 FAILED! => {
    "changed": false,
    "current": {},
    "invocation": {
        "module_args": {
            "bd": "BD-101",
            "dhcp_option_policy": null,
            "dhcp_policy": "DHCP_RELAY_HSM_POLICY1",
            "host": "10.48.187.200",
            "login_domain": null,
            "name": "DHCP_RELAY_HSM_POLICY1",
            "output_level": "normal",
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "port": null,
            "schema": "HSM_TN_NDO_423K_VRF",
            "state": "present",
            "template": "DC5",
            "timeout": null,
            "use_proxy": null,
            "use_ssl": null,
            "username": "admin",
            "validate_certs": false,
            "version": 1
        }
    },
    "msg": "DHCP policy 'DHCP_RELAY_HSM_POLICY1' does not exist"
}

### Expected Behavior

* DHCP relay policy must be associated correctly.

### Actual Behavior

* error at playbook execution

### Playbook tasks to Reproduce

(ansible-ndo) :~/ansible-ndo/ndo-playbooks$ cat 0001_add_bd_dhcp_config.yml
---
- name: =====  NDO 4.2(3k) / ND 3.0(1i) Testing  =====
  hosts: ndo
  connection: local
  gather_facts: false

  tasks:

    - name: =====  Add a new DHCP policy under BD  =====
      cisco.mso.mso_schema_template_bd_dhcp_policy:
        host: "{{ ansible_host }}"
        username: "{{ username }}"
        password: "{{ password }}"
        validate_certs: False
        schema: "HSM_TN_NDO_423K_VRF"
        template: "DC5"
        bd: BD-101
        name: DHCP_RELAY_HSM_POLICY1
        version: 1
        state: present

### References

* #0000

### PR DESCRIPTION
**Pull request by akinross**
with ndo_schema_template_bd_dhcp_policy

fixes #495

### CODE REVIEW

File: tests/integration/targets/ndo_schema_template_bd_dhcp_policy/tasks/main.yml [hunk 1]
```diff
@@ -0,0 +1,409 @@
...
+      ignore_errors: true
+      register: err_relay_policy_not_exist
+    
+    - name: Assert errors
+      assert:
+        that:
+          - err_relay_policy_without_template is failed
+          - err_relay_policy_without_template.msg == "state is present but all of the following are missing{{":"}} dhcp_relay_policy"
+          - err_relay_policy_not_exist is failed
+          - err_relay_policy_not_exist.msg == "Provided DHCP Relay Policy with '[KVPair(key='name', value='fake_relay_policy')]' not matching existing object(s){{":"}} ansible_test_relay_policy_2"
```
Comment (sajagana) [inline tests/integration/targets/ndo_schema_template_bd_dhcp_policy/tasks/main.yml:348]: Can we use `'[KVPair(key='name', value='fake_relay_policy')]'` in the error message?

**Review by sajagana — CHANGES_REQUESTED** [review]

File: plugins/modules/ndo_schema_template_bd_dhcp_policy.py [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+            ["state", "present", ["dhcp_relay_policy"]],
+        ],
+        required_together=[
+            ["dhcp_relay_policy", "dhcp_relay_policy_template"],
+        ],
+    )
+
+    schema = module.params.get("schema")
+    template = module.params.get("template").replace(" ", "")
+    bd = module.params.get("bd")
```
[the remark this answers is not in the document]
Comment (akinross) [reply plugins/modules/ndo_schema_template_bd_dhcp_policy.py:158]: kept this consistent with mso_schema_template_bd_dhcp_policy

File: tests/integration/targets/ndo_schema_template_bd_dhcp_policy/tasks/main.yml (hunk 1 above)
Comment (akinross) [reply tests/integration/targets/ndo_schema_template_bd_dhcp_policy/tasks/main.yml:348]: Is there a reason why we cannot?

**Review by sajagana — APPROVED** [review]
LGTM!

File: plugins/modules/ndo_schema_template_bd_dhcp_policy.py [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+        state=dict(type="str", default="present", choices=["absent", "present", "query"]),
+    )
+
+    module = AnsibleModule(
+        argument_spec=argument_spec,
+        supports_check_mode=True,
+        required_if=[
+            ["state", "absent", ["dhcp_relay_policy"]],
+            ["state", "present", ["dhcp_relay_policy"]],
+        ],
```
Comment (gmicol) [inline plugins/modules/ndo_schema_template_bd_dhcp_policy.py:131]: Just a quick suggestion for dhcp_option_policy.

```suggestion
        ],
        required_by={
           ["dhcp_option_policy": "dhcp_relay_policy",
        },
```

**Review by gmicol — CHANGES_REQUESTED** [review]

**Review by sajagana — APPROVED** [review]
LGTM!

**Review by anvitha-jain — DISMISSED** [review]

File: plugins/modules/ndo_schema_template_bd_dhcp_policy.py [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+    type: str
+  state:
+    description:
+    - Use C(present) or C(absent) for adding or removing.
+    - Use C(query) for listing an object or multiple objects.
+    type: str
+    choices: [ absent, present, query ]
+    default: present
+notes:
+- This module can only be used on versions of MSO that are 4.x or greater.
```
Comment (shrsr) [inline plugins/modules/ndo_schema_template_bd_dhcp_policy.py:54]: - This module can only be used on versions of NDO that are 4.x or greater.

File: plugins/modules/ndo_schema_template_bd_dhcp_policy.py [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+                dhcp_relay_policy["dhcpOptionLabel"]["name"] = get_dhcp_option_label_name(mso, dhcp_option_label_ref)
+
+    if state == "query":
+        mso.existing = existing_dhcp_relay_policy
+        mso.exit_json()
+
+    ops = []
+    mso.previous = existing_dhcp_relay_policy
+
+    if state == "absent":
```
Comment (shrsr) [inline plugins/modules/ndo_schema_template_bd_dhcp_policy.py:178]:
```
 if state == "absent" and existing_dhcp_relay_policy:
    mso.sent = mso.existing = {}
    ops.append(dict(op="remove", path=dhcp_labels_path))
```

File: plugins/modules/ndo_schema_template_bd_dhcp_policy.py [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+        if existing_dhcp_relay_policy:
+            mso.sent = mso.existing = {}
+            ops.append(dict(op="remove", path=dhcp_labels_path))
+
+    if state == "present":
```
Comment (shrsr) [inline plugins/modules/ndo_schema_template_bd_dhcp_policy.py:183]: Would be better for readability->
elif state == "present":

**Review by shrsr — CHANGES_REQUESTED** [review]

File: plugins/modules/ndo_schema_template_bd_dhcp_policy.py (hunk 4 above)
Comment (akinross) [reply plugins/modules/ndo_schema_template_bd_dhcp_policy.py:54]: ok

File: plugins/modules/ndo_schema_template_bd_dhcp_policy.py (hunk 5 above)
Comment (akinross) [reply plugins/modules/ndo_schema_template_bd_dhcp_policy.py:178]: ok and removed redundant empty setting of mso.sent and mso.existing

File: plugins/modules/ndo_schema_template_bd_dhcp_policy.py (hunk 6 above)
Comment (akinross) [reply plugins/modules/ndo_schema_template_bd_dhcp_policy.py:183]: ok

**Review by sajagana — APPROVED** [review]
LGTM!

**Review by shrsr — APPROVED** [review]
LGTM

**Review by gmicol — DISMISSED** [review]
LGTM

**Review by samiib — DISMISSED** [review]
LGTM

File: plugins/module_utils/schema.py [hunk 7]
```diff
@@ -112,6 +112,21 @@ def set_template_bd(self, bd, fail_module=True):
             self.mso.fail_json(msg=msg)
         self.schema_objects["template_bd"] = match
 
+    def set_template_bd_dhcp_relay_policy(self, relay_policy_ref, fail_module=True):
+        """
+        Get template bridge domain item that matches the name of a bd.
```
Comment (lhercot) [inline plugins/module_utils/schema.py:117]:
```suggestion
        Get template bridge domain dhcp relay policy item that matches the name of a bd.
```

File: plugins/modules/ndo_schema_template_bd_dhcp_policy.py [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+    type: str
+  state:
+    description:
+    - Use C(present) or C(absent) for adding or removing.
+    - Use C(query) for listing an object or multiple objects.
+    type: str
+    choices: [ absent, present, query ]
+    default: present
+notes:
+- This module can only be used on versions of NDO that are 4.x or greater.
```
Comment (lhercot) [inline plugins/modules/ndo_schema_template_bd_dhcp_policy.py:54]:
```suggestion
- This module can only be used on versions of NDO that are 4.1 or greater.
```
We don't support 4.0 so at least 4.1 here.

Comment (lhercot) [reply plugins/modules/ndo_schema_template_bd_dhcp_policy.py:54]: We should also add a message on the older module that it shouldn't be used up to a certain version and to use this one for NDO 4.1+

**Review by lhercot — CHANGES_REQUESTED** [review]

**Review by lhercot — APPROVED** [review]
LGTM

**Pull request merged**